src/ronin/migration/LegacyTokenMigrator.sol

```diff
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.27;
+
+import { Initializable } from "@openzeppelin/contracts/proxy/utils/Initializable.sol";
+import { AccessControlEnumerable } from "@openzeppelin/contracts/access/AccessControlEnumerable.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { IERC20Metadata } from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+interface IERC20MintBurn {
+  function mint(address to, uint256 amount) external;
+
+  function burn(
+    uint256 amount
+  ) external;
+
+  function burnFrom(address account, uint256 amount) external;
+}
+
+contract LegacyTokenMigrator is Initializable, AccessControlEnumerable {
+  using SafeERC20 for IERC20;
+
+  mapping(address legacyToken => address newToken) internal _tokenMap;
+
+  error MintTokenFailed(address token, uint256 amount);
+  error NullAmount();
+  error TokenNotMapped();
+  error LengthMismatch();
+  error DecimalsMismatch(uint8 legacyDecimals, uint8 newDecimals);
+
+  event TokenMapped(address indexed by, address indexed legacyToken, address indexed newToken);
+  event Converted(address indexed user, address indexed fromToken, address indexed toToken, uint256 amount);
+  event LegacyTokenBurned(address indexed by, address indexed token, uint256 amount);
+  event LegacyTokenLocked(address indexed by, address indexed token, uint256 amount);
+
+  constructor() {
+    _disableInitializers();
+  }
+
+  function initialize(address admin, address[] calldata legacyTokens, address[] calldata newTokens) external initializer {
```

Comment:
Should check with legal to see if the upgradable pattern is allowed, since this contract holds funds